firedrake/preconditioners/asm.py

```diff
@@ -36,10 +36,24 @@ def initialize(self, pc):
         # Create new PC object as ASM type and set index sets for patches
         asmpc = PETSc.PC().create(comm=pc.comm)
         asmpc.incrementTabLevel(1, parent=pc)
-        asmpc.setOptionsPrefix(self.prefix + "_sub")
+        asmpc.setOptionsPrefix(self.prefix + "sub_")
         asmpc.setOperators(*pc.getOperators())
         asmpc.setType(asmpc.Type.ASM)
         asmpc.setASMLocalSubdomains(len(ises), ises)
+
+        # Set default solver parameters
+        asmpc.setASMType(PETSc.PC.ASMType.BASIC)
+        opts = PETSc.Options(asmpc.getOptionsPrefix())
+        if "sub_pc_type" not in opts:
+            opts["sub_pc_type"] = "lu"
+        if "sub_pc_factor_shift_type" not in opts:
+            opts["sub_pc_factor_shift_type"] = "NONE"
```

Comment:
Can simplify these to `opts.setdefault(key, defaultvalue)`

Reply:
Thanks, fixed.

Reply:
This fix broke things, failing with `AttributeError: 'petsc4py.PETSc.Options' object has no attribute 'setdefault'`.  Looking at the `petsc4py` source, that seems both true and not just a matter of adjusting `setdefault`.  Should we revert or is there a better way?

Reply:
Ah sorry, I had not twigged that opts here is just the Options database which isn't fully like a dict. Just revert.